pkg/services/notary/notary.go

```diff
@@ -282,8 +289,14 @@ func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
 	// the copy.
 	r.fallbacks = append(r.fallbacks, payload.FallbackTransaction.Copy())
 	if exists && r.isMainCompleted() || validationErr != nil {
+		r.lock.Unlock()
 		return
 	}
+
+	// @roman-khimov, I tried to work with r.copy() here in order not to hold r.lock during
+	// witness verification. However, the race is possible in the end of this method if main
+	// transaction will be completed by another thread, which leads to the situation when
+	// two main transactions with different set of witnesses will be submitted.
```

Comment:
Probably this can be restructured into pure verification part (concurrent) and merging the result into the final tx (consecutive). It at the same time won't be that effective in terms of optimizations. We're talking about request spikes on the order of hundreds/thousands. Each separate main transaction is likely to have just 5/7 requests, so once we're to allow different transactions to be processed (mostly) concurrently we'll solve ~98% of the problem.